staging/src/k8s.io/apiserver/pkg/cel/quantity.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+var (
+	QuantityObject    = decls.NewObjectType("kubernetes.Quantity")
+	quantityTypeValue = types.NewTypeValue("kubernetes.Quantity")
+	QuantityType      = cel.ObjectType("kubernetes.Quantity")
+)
+
+// Quantity provdes a CEL representation of a resource.Quantity
+type Quantity struct {
+	*resource.Quantity
+}
+
+func (d Quantity) ConvertToNative(typeDesc reflect.Type) (interface{}, error) {
+	if reflect.TypeOf(d.Quantity).AssignableTo(typeDesc) {
+		return d.Quantity, nil
+	}
+	if reflect.TypeOf("").AssignableTo(typeDesc) {
+		return d.Quantity.String(), nil
+	}
+	return nil, fmt.Errorf("type conversion error from 'Quantity' to '%v'", typeDesc)
+}
+
+func (d Quantity) ConvertToType(typeVal ref.Type) ref.Val {
+	switch typeVal {
+	case typeValue:
+		return d
+	case types.TypeType:
+		return quantityTypeValue
+	default:
+		return types.NewErr("type conversion error from '%s' to '%s'", quantityTypeValue, typeVal)
+	}
+}
+
+func (d Quantity) Equal(other ref.Val) ref.Val {
+	otherDur, ok := other.(Quantity)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(other)
+	}
+	return types.Bool(d.Quantity.Equal(*otherDur.Quantity))
+}
```

Comment:
If we don't support comparison operations for quantity, maybe include rationale for why near this declaration?  The presence of the `Equal` operator made me start wondering if we should support the comparison trait.

Reply:
I was looking into how to overload the `>` , `<`, etc operators for (Quantity, Quantity) tuples, but not sure where the Trait is documented. (do not see it appearing [here](https://pkg.go.dev/github.com/google/cel-go/common/types/traits)). I think it'd be good to add.

Reply:
Had an attempt at implementing this. Turns out the traits I linked are not respected when using a type defined by `types.NewObjectValue` to define library functions. @jiahuif looked into it some more and may have more background about technical restrictions. For now I will revert to using functions rather than the operators with a note about it

Reply:
Sounds good. If it ends up being a major problem to implement let me know and we can decide what functions to offer.

Reply:
Sorry if I did not explain well - I don't think we can implement the traits on the new type. The `cel-go` api for providing functions through the `cel.Function` EnvOpt does not support custom object types with traits, so we are offering functions instead.

Specifically, `cel.Function` requires types to be specified by a `*cel.Type` instance, and it is not possible to construct one of these which expose support for traits. I believe @jiahuif experimented with it and found that it was as simple as simply exposing a constructor.

Reply:
I think what you have is OK. Thanks for digging in and figuring out what is supported.